internal/flowcontrol/writebuffer.go

```diff
@@ -113,15 +115,14 @@ func (w *ResourceSliceWriteBuffer) processQueueItem(ctx context.Context) bool {
 	if len(updates) > max {
 		w.state[sliceNSN] = updates[max:]
 		updates = updates[:max]
-	} else {
-		delete(w.insertionTime, sliceNSN)
 	}
 	w.mut.Unlock()
 
 	// We only forget the rate limit once the update queue for this slice is empty.
 	// So the first write is fast, but a steady stream of writes will be throttled exponentially.
 	if len(updates) == 0 {
 		w.queue.Forget(item)
+		delete(w.insertionTime, sliceNSN)
```

Comment:
should this delete run in mutex lock also? the len(updates) might be different outside the mutex block

Reply:
good catch! Fixed and confirmed that the tests pass with the race detector enabled